docs/readmes/orc8r/architecture_modularity.md

```diff
@@ -0,0 +1,99 @@
+
+Orchestrator provides a discrete set of extensions for injecting functionality into the core services. These extensions (i.e. hooks)
+implement a per-extension gRPC servicer, where core Orchestrator services will handle making calls over-the-network to the servicer at
+the appropriate time. Each service in the deployment can implement any of the extensions, and the core services will automatically
+handle calling the extension implementation. Discovery is handled by assigning Kubernetes labels and annotations.
+
+There are two categories of extensions
+
+- *Producer* extensions export data to core services upon request
+- *Consumer* extensions receive data from core services
+
+### Supported extensions
+
+We currently support 7 extensions
+
+- **Analytics collector**
+    - Calculate, write, and return new metrics derived from existing metrics
+    - Producer, aggregated by the *analytics* service
+    - Label: `orc8r.io/analytics_collector`
+- **Mconfig builder**
+    - Define configuration (mconfigs) for gateway services
+    - Producer, aggregated by the *configurator* service
+    - Label: `orc8r.io/mconfig_builder`
+- **Metrics exporter**
+    - Export metrics to chosen data sink
+    - Consumer, pushed from the *metricsd* service
+    - Label: `orc8r.io/metrics_exporter`
+- **Obsidian handler** (REST API)
+    - Define REST API endpoints
+    - Producer/consumer, proxied by the *obsidian* service
+    - Label: `orc8r.io/obsidian_handlers`
+    - Annotations
+        - `orc8r.io/obsidian_handlers_path_prefixes` determine which endpoints to proxy
+- **State indexer**
+    - Receive reported state for processing and/or indexing under new primary key
+    - Consumer, pushed from the *state* service
+    - Label: `orc8r.io/state_indexer`
+    - Annotations
+        - `orc8r.io/state_indexer_version` reindex all state when indexer version is incremented
+        - `orc8r.io/state_indexer_types` which types of state to send to the indexer
+- **Stream provider**
+    - Push arbitrary objects to gateway services subscribed to the stream
+    - Producer, aggregated by the *streamer* service
+    - Label: `orc8r.io/stream_provider`
+    - Annotations
+        - `orc8r.io/stream_provider_streams` which streams the provider exposes
+- **Swagger specifier**
+    - Define the accompanying REST API endpoints by exposing a Swagger (OpenAPI) specification
+    - Producer, aggregated by the *obsidian* service
+    - Label: `orc8r.io/swagger_spec`
+
+### Example extension
+
+The interface between core service and extension implementation is defined per extension. Beyond that interface, functionality depends on
+the particular extension. For a concrete example, consider the state indexing codepath
+
+![State indexing codepath](assets/orc8r/state_indexing_codepath.png)
+
+## Extending Orchestrator
+
+### Implement extension
+
+You'll need to define a service executable which implements the gRPC servicer definition for one of the above extensions. Then,
```

Comment:
Not immediate requirement. It might be nice if we can add some example service here, which someone can copy/paste.